fusion/FSL.md

```diff
@@ -0,0 +1,1220 @@
+
+Comments use '# ' (hash/space) for single line comments or '###' as 
+the delimiter for comment blocks. Comments can be embedded in other comments.
+
+```
+#  ### commenting out
+#  my multi line
+#  comment by
+#  prefixing the '#'
+#  end of ML comment ###
+```
+
+Mixing block and line comments is supported, the first comment token
+takes precedence. This practice is not recommended.
+
+```
+#  ### commenting out
+#  my multi line
+   this line will be processed
+#  prefixing the '#'
+#  end of ML comment ###
+```
+
+In this case the block comment start and stop markers are not processed 
+because they are delimited by line comments.  This is supported but 'this
+line will be processed' will be tokenized and tested for legal grammar. 
+
+## Preprocessor Support
+
+The FSL parser uses a pre-processor with features similar to that found
+in the C pre-processor.  Within the constraints of the FSL language most 
+pre-processor operations found in C will be available in FSL with slight
+changes for FSL syntax choices.
+ 
+Note that the typical # indicator, as in #include, is replaced with
+'`' as in `include.
+
+### Conditional Compilation
+
+FSL supports a common set of conditional compilation directives, but 
+with the typical # replaced by back-tick.
+
+```
+`ifdef X
+`ifndef X
+`if X == Y
+
+`endif
+`endif
+`endif
+```
+
+### Includes
+
+The include keyword declares external files. The parser handles 
+included files by expanding them into the current position in
+the current translation unit. Include recursion is illegal.
+
+The syntax is:
+
+`include myfile.fh
+
+The .fh extension is preferred but not enforced by the parser. Quotes
+around the file name are not used.
+
+The single inclusion pragma is declared as:
+
+`once
+
+This pragma must be expressed before any other non-comment statement.
+
+## Transform Structure
+
+FSL transform specifications are read at runtime. The FSL file 
+list is passed to the Fusion API constructor. The API invokes the
+FSL parser.
+
+A transform's internal structure consists of a number of clauses,
+the sequence, constraints, and conversion clauses, along with a number
+elements in the prolog which specify the context of the fusion operation,
+ISA, microarchitecture and hooks into the performance model.
+
+## Transform Specification 
+
+The transform statement is a top level structure.  Each transform statement
+is named. The name must be unique across all translation units.
+
+The generic form is shown below. The order of the sections and clauses
+is arbitrary. What is shown is the recommended convention.
+
+```
+transform <name> {
+  <prolog section>
+  <sequence clause>
+  <constraints clause>
+  <conversion clause>
+}
+```
+
+## Prolog Elements
+
+Prolog is not a keyword. The prolog elements provide the necessary 
+context for the transform operation and they are commonly referred to 
+as a group.  The common elements are shown:
+
+```
+  isa    myIsa
+  uarch  myImplementation
+  ioput  iop1
+```
+
```

Comment:
Can prolog elements be considered as a form of constraint? If so, can this be made part of constraints?

Reply:
I think of them more as context although that distinction is not always sharp.  Having them where they are in the syntax makes their scope explicit, they are used by the other clauses, it also makes the library style a consistent extension of this style, the prolog elements can be referenced as entities rather than a sub-entity of a named constraint object.